### ISSUE TITLE
Data race on sshContext

### ISSUE THREAD

**Issue by prologic** [gliderlabs/ssh#160]
came across this data race:

```#!sh
==================
WARNING: DATA RACE
Write at 0x00c0000988d0 by goroutine 14:
  github.com/gliderlabs/ssh.(*sshContext).SetValue()
      /Users/prologic/go/pkg/mod/github.com/gliderlabs/ssh@v0.3.3/context.go:123 +0xa6
  github.com/gliderlabs/ssh.(*Server).config.func2()
      /Users/prologic/go/pkg/mod/github.com/gliderlabs/ssh@v0.3.3/server.go:150 +0x18b
  golang.org/x/crypto/ssh.(*connection).serverAuthenticate()
      /Users/prologic/go/pkg/mod/golang.org/x/crypto@v0.0.0-20210616213533-5ff15b29337e/ssh/server.go:523 +0x2799
  golang.org/x/crypto/ssh.(*connection).serverHandshake()
      /Users/prologic/go/pkg/mod/golang.org/x/crypto@v0.0.0-20210616213533-5ff15b29337e/ssh/server.go:277 +0xb78
  golang.org/x/crypto/ssh.NewServerConn()
      /Users/prologic/go/pkg/mod/golang.org/x/crypto@v0.0.0-20210616213533-5ff15b29337e/ssh/server.go:206 +0x2ab
  github.com/gliderlabs/ssh.(*Server).HandleConn()
      /Users/prologic/go/pkg/mod/github.com/gliderlabs/ssh@v0.3.3/server.go:281 +0x2af

Previous read at 0x00c0000988d0 by goroutine 39:
  github.com/gliderlabs/ssh.(*sshContext).Done()
      <autogenerated>:1 +0x48
  net/http.http2awaitRequestCancel()
      /usr/local/Cellar/go/1.16.6/libexec/src/net/http/h2_bundle.go:6823 +0x117
  net/http.(*http2clientStream).awaitRequestCancel()
      /usr/local/Cellar/go/1.16.6/libexec/src/net/http/h2_bundle.go:6846 +0x64
```

**Comment by belak** [comment]
Do you have sample code, or was this when running tests? Unfortunately, unless we happen to run across the same race, there's no way to debug this.

Also, I just noticed that this is coming from inside an http handler, so there's a larger likelihood that it's an issue with your code somewhere.

**Comment by belak** [comment]
We haven't gotten a response in almost a week, so I'm closing this for now. Feel free to comment with more information or file a new issue and we'll do our best to address this.

**Comment by prologic** [comment]
> We haven't gotten a response in almost a week, so I'm closing this for now. Feel free to comment with more information or file a new issue and we'll do our best to address this.

Oh I'm so sorry! This issue got lost in the "ether" for me. Can we please reopen it, and I'll get you a reproducer ASAP 👌

**Comment by prologic** [comment]
I cannot reproduce this anymore. I _may_ have filed it in error, but nearly a year ago, so I'm not completely sure. Sorry! I ran `go test -race .` on this repo at `v0.3.3` and the latester `master` -- no problems 👌

**Comment by aymanbagabas** [comment]
This issue still persists :/

**Comment by belak** [comment]
@aymanbagabas Where are you seeing this data race? Do you have a way of reproducing it?

**Comment by aymanbagabas** [comment]
@belak I don't have a MRE right now, but I'm noticing this issue here https://github.com/charmbracelet/soft-serve/actions/runs/5535584234/jobs/10102073239?pr=337#step:4:279

**Comment by aymanbagabas** [comment]
@belak This patch works for me https://github.com/gliderlabs/ssh/pull/209

**Comment by wxiaoguang** [comment]
Quote:

> The fix is incomplete.
> Every call to sshCtx.Context (including `Err()` and others) all causes Data Race.

**Comment by wxiaoguang** [comment]
IMO the complete fix could be like this: Fix data race by using atomic value #211
